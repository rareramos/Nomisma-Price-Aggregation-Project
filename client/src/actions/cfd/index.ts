import { types } from 'actions/types';
import { emptyActionCreator, payloadActionCreator } from 'actions/action-creator-factories';
import {
  IQuasiLive,
  ICfdFormState,
  IMarketsDataState,
  IUpdateableMarketData,
  ISocketState,
} from 'types/reducers/cfd';
import { IHash } from 'types/utils';

export const onFormChange = payloadActionCreator<
types.CFD_FORM_CHANGE,
ICfdFormState
>(types.CFD_FORM_CHANGE);

export const startQuasiLiveFetch = emptyActionCreator(types.CFD_QUASILIVE_START);
export const fetchQuasiLiveData = emptyActionCreator(types.CFD_QUASILIVE_FETCH);

export const saveQuasiLiveData = payloadActionCreator<
types.CFD_QUASILIVE_SAVE,
Array<IQuasiLive>
>(types.CFD_QUASILIVE_SAVE);

export const addAvailableMarkets = payloadActionCreator<
types.ADD_AVAILABLE_MARKETS,
Array<string>
>(types.ADD_AVAILABLE_MARKETS);

export const onMarketSelect = payloadActionCreator<
types.MARKET_SELECTED,
string
>(types.MARKET_SELECTED);

export const addBulkMarketData = payloadActionCreator<
types.ADD_BULK_MARKET_DATA,
IMarketsDataState
>(types.ADD_BULK_MARKET_DATA);

export const fetchAllMarketData = payloadActionCreator<
types.MARKET_DATA_ALL,
IMarketsDataState
>(types.MARKET_DATA_ALL);

export const onUpdateMarketData = payloadActionCreator<
types.MARKET_DATA_UPDATED,
Array<IHash<IUpdateableMarketData>>
>(types.MARKET_DATA_UPDATED);

export const addCFDSocket = payloadActionCreator<
types.ADD_CFD_SOCKET,
Partial<ISocketState>
>(types.ADD_CFD_SOCKET);

export const marketUpdateParameters = emptyActionCreator(types.MARKET_UPDATE_PARAMETERS);
export const toggleTailoredView = emptyActionCreator(types.CFD_TAILORED_VIEW_TOGGLE);
export const toggleCFDModal = emptyActionCreator(types.CFD_MODAL_VIEW_TOGGLE);
export const onCFDModalTabChange = emptyActionCreator(types.CFD_MODAL_TAB_CHANGE);
export const onCFDModalClose = emptyActionCreator(types.CFD_MODAL_CLOSE);
