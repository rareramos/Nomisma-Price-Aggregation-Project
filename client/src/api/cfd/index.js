export * from './socket-setup';
export * from './quasi-live';
