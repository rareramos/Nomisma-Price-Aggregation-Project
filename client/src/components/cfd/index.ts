export { Cfd } from './cfd-container';
