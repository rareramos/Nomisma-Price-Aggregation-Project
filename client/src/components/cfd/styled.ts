import styled from 'styled-components';
import { Grid } from '@nomisma/nomisma-ui/grid';
import { SimpleTable } from '@nomisma/nomisma-ui/table/simple-table';
import { ccpTableStyle } from '@nomisma/nomisma-ui/table/simple-table/styles';

export const Styled = styled.div`
  text-align: center;

  table {
    text-align: left;
  }

  label {
    text-align: left;
  }
`;

export const Title = styled.h5`
  margin: 2rem 0 0;
  text-align: left;
`;

export const GridContainer = styled(Grid)`
  margin: 1rem 0 2rem;
`;

export const Link = styled.a.attrs({
  href: props => props.href,
  target: props => props.target,
})`
  cursor: pointer;
  justify-self: end;
`;

export const StyledTable = styled(SimpleTable)(ccpTableStyle, {
  '.thead': {
    backgroundColor: 'inherit',
    marginTop: '3em',
  },
  '.th': {
    textAlign: 'center',
  },
  '.tbody': {
    backgroundColor: 'white',
    '.tr': {
      borderBottom: '1px solid var(--bgLight)',
    },
    '.tr:nth-of-type(2n)': {
      backgroundColor: 'inherit',
    },
    '.tr:hover': {
      backgroundColor: 'var(--bgLight)',
    },
  },
});

export const ProviderWrapper = styled.div`
  align-items: center;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-around;
`;

export const StyledText = styled.p`
  margin: 0;
  padding: 0;
`;
