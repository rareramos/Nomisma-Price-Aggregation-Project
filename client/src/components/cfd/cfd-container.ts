import { connect } from 'react-redux';
import { ICfdDispatchProps } from 'types/components/cfd';
import { Cfd as CfdComp } from './cfd';
import {
  startQuasiLiveFetch,
} from '../../actions/cfd';

const mapDispatchToProps : ICfdDispatchProps = { startQuasiLiveFetch };

export const Cfd = connect(
  null,
  mapDispatchToProps,
)(CfdComp);
