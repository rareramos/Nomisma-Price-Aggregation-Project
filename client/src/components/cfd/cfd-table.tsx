import React, { ReactElement } from 'react';
import { IconQtum } from '@nomisma/nomisma-ui/token-icons';
import { ICfdTableProps } from 'types/components/cfd';
import {
  Styled,
  StyledTable,
  ProviderWrapper,
  StyledText,
} from './styled';

export const CfdTable = ({
  cfdHeader,
  cfdKey,
  cfdBody,
  uniqueProviderNames,
} : ICfdTableProps) : ReactElement => {
  function resolveBodyData() : Array<Array<ReactElement>> {
    return cfdBody.length > 0 && cfdBody.map((provider : Array<string>) => provider.map((item : string) => {
      if (uniqueProviderNames.filter((providerName : string) => providerName === item).length > 0) {
        return (
          <ProviderWrapper>
            <IconQtum
              fontSize="16px"
            />
            <StyledText>
              { item }
            </StyledText>
          </ProviderWrapper>
        );
      }
      return (
        <StyledText>
          { item }
        </StyledText>
      );
    }));
  }

  return (
    <Styled>
      { cfdBody.length > 0 && (
        <StyledTable
          header={cfdHeader}
          body={resolveBodyData()}
          keys={cfdKey}
        />
      ) }
    </Styled>
  );
};
