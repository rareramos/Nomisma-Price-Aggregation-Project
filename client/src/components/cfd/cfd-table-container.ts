import { connect } from 'react-redux';
import { ICfdTableMapProps } from 'types/components/cfd';
import { ICfdState } from 'reducers/cfd';
import { CfdTable as CfdTableComp } from './cfd-table';
import {
  selectCfdColumnsFromState,
  selectCfdDisplayTableRecordsFromState,
  selectCfdHeaderFromState,
  selectCfdKeyFromState,
  selectCfdBodyFromState,
  selectUniqueProvidersFromState,
} from '../../selectors/cfd';

const mapStateToProps = (state : ICfdState) : ICfdTableMapProps => ({
  cfdHeader: selectCfdHeaderFromState(state),
  cfdKey: selectCfdKeyFromState(state),
  cfdBody: selectCfdBodyFromState(state),
  tableData: selectCfdDisplayTableRecordsFromState(state),
  columns: selectCfdColumnsFromState(state),
  uniqueProviderNames: selectUniqueProvidersFromState(state),
});

const mapDispatchToProps = null;

const options = {
  areStatesEqual(next, prev) {
    return prev.cfd.marketData === next.cfd.marketData && prev.cfd.quasiLive === next.cfd.quasiLive;
  },
};

export const CfdTable = connect(
  mapStateToProps,
  mapDispatchToProps,
  null,
  options,
)(CfdTableComp);
