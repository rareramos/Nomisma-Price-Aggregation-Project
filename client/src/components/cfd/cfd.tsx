import React, { Component, ReactNode } from 'react';
import { Grid } from '@nomisma/nomisma-ui/grid';
import { ICfdProps } from 'types/components/cfd';
import { CfdTable } from './cfd-table-container';
import { Preferences } from './components/preferences/preferences-container';

export class Cfd extends Component<ICfdProps> {
  public componentDidMount() : void {
    const { startQuasiLiveFetch } = this.props;
    startQuasiLiveFetch();
  }

  public render() : ReactNode {
    return (
      <Grid
        column="1fr 10fr 1fr"
      >
        <div />
        <div>
          <Preferences />
          <CfdTable />
        </div>
        <div />
      </Grid>
    );
  }
}
