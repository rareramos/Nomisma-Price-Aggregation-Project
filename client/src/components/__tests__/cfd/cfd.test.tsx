import React from 'react';
import { shallow } from 'enzyme';
import {
  ICfdProps,
  ICfdTableProps,
} from 'types/components/cfd';

import { Cfd } from 'components/cfd/cfd';
import { CfdTable } from 'components/cfd/cfd-table';
import { cfdColumns } from 'utils';

const CfdProps : ICfdProps = {
  startQuasiLiveFetch: jest.fn,
};

const CfdTableProps : ICfdTableProps = {
  tableData: [],
  columns: cfdColumns,
  uniqueProviderNames: [],
  cfdHeader: [],
  cfdKey: [],
  cfdBody: [],
};

describe('CFD pipeline snapshot test', () => {
  it('matches cfd snapshot', () => {
    const wrapper = shallow(<Cfd {...CfdProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches cfd table snapshot', () => {
    const wrapper = shallow(<CfdTable {...CfdTableProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
