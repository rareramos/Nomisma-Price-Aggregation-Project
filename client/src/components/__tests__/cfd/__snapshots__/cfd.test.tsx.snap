// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFD pipeline snapshot test matches cfd snapshot 1`] = `
<styled.div
  column="1fr 10fr 1fr"
>
  <div />
  <div>
    <Connect(Preferences) />
    <Connect(CfdTable) />
  </div>
  <div />
</styled.div>
`;

exports[`CFD pipeline snapshot test matches cfd table snapshot 1`] = `<styled.div />`;
